usb_c: sink: unit: Add unit test for USB-C Sink testing

sambhurst opened the pull request:
Add unit tests for the USB-C subsystem. This PR only tests that the subsystem can connect as a Sink and negotiate an explicit contract. Following PR's will add additional Sink tests and any new USB-C features will included unit tests.

sambhurst committed:
usb_c: sink: unit: Add unit test for USB-C Sink testing

Add USB-C Sink testing. This change only tests that
a the Sink subsystem can detected a Source and
negotiate an explicit contract.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb_c: Fix usbc_policy enum type

Correct the usbc_policy enum type of dpm_pwr_change_notify. The correct
type is enum usbc_policy_notify_t

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb_c: Fix message check of PD Data message

Use received_data_message to check a data message instead of
incorrectly using received_control_message.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

desowin commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -345,7 +345,9 @@ void pe_set_state(const struct device *dev, const enum usbc_pe_state state)
 {
 	struct usbc_port_data *data = dev->data;
 
+#if !defined(CONFIG_TEST)
```
Why remove the assert when `CONFIG_TEST` is enabled? The point behind asserts is to alert programmer about conditions that must never be false but do become false due to incorrect code (e.g. invalid API usage).

sambhurst replied:
I agree that this is not the correct thing to do but twister was complaining:
```
/home/sam/Projects/github/zephyrproject/zephyr/include/zephyr/sys/util.h:109:45: error: invalid application of 'sizeof' to incomplete type 'const struct smf_state[]'
  109 |         ((size_t) (IS_ARRAY(array) + (sizeof(array) / sizeof((array)[0]))))
      |                                             ^
/home/sam/Projects/github/zephyrproject/zephyr/include/zephyr/sys/__assert.h:116:23: note: in definition of macro '__ASSERT'
  116 |                 if (!(test)) {                                            \
      |                       ^~~~
/home/sam/Projects/github/zephyrproject/zephyr/subsys/usb/usb_c/usbc_pe_common.c:348:26: note: in expansion of macro 'ARRAY_SIZE'
  348 |         __ASSERT(state < ARRAY_SIZE(pe_states), "invalid pe_state %d", state);
      |                          ^~~~~~~~~~
```
But the build works fine.

alevkoy replied:
How odd. It seems like `const struct smf_state[]` should be completely defined regardless of `CONFIG_TEST`.

Is that newly happening in this PR?

desowin commented on subsys/usb/usb_c/usbc_prl.c:
```diff
@@ -380,7 +380,9 @@ static void prl_tx_set_state(const struct device *dev, const enum usbc_prl_tx_st
 	struct usbc_port_data *data = dev->data;
 	struct protocol_layer_tx_t *prl_tx = data->prl_tx;
 
+#if !defined(CONFIG_TEST)
```
Why remove the assert when `CONFIG_TEST` is enabled? The point behind asserts is to alert programmer about conditions that must never be false but do become false due to incorrect code (e.g. invalid API usage).

desowin commented on subsys/usb/usb_c/usbc_prl.c:
```diff
@@ -392,7 +394,9 @@ static void prl_hr_set_state(const struct device *dev, const enum usbc_prl_hr_st
 	struct usbc_port_data *data = dev->data;
 	struct protocol_hard_reset_t *prl_hr = data->prl_hr;
 
+#if !defined(CONFIG_TEST)
```
Why remove the assert when `CONFIG_TEST` is enabled? The point behind asserts is to alert programmer about conditions that must never be false but do become false due to incorrect code (e.g. invalid API usage).

sambhurst replied:
I agree that this is not the correct thing to do but twister was complaining:
```
/home/sam/Projects/github/zephyrproject/zephyr/include/zephyr/sys/util.h:109:45: error: invalid application of 'sizeof' to incomplete type 'const struct smf_state[]'
  109 |         ((size_t) (IS_ARRAY(array) + (sizeof(array) / sizeof((array)[0]))))
      |                                             ^
/home/sam/Projects/github/zephyrproject/zephyr/include/zephyr/sys/__assert.h:116:23: note: in definition of macro '__ASSERT'
  116 |                 if (!(test)) {                                            \
      |                       ^~~~
/home/sam/Projects/github/zephyrproject/zephyr/subsys/usb/usb_c/usbc_prl.c:395:26: note: in expansion of macro 'ARRAY_SIZE'
  395 |         __ASSERT(state < ARRAY_SIZE(prl_hr_states), "invalid prl_hr_state %d", state);
```
But the build works fine.

desowin commented on subsys/usb/usb_c/usbc_tc_common.c:
```diff
@@ -127,7 +127,9 @@ void tc_set_state(const struct device *dev, const enum tc_state_t state)
 	struct usbc_port_data *data = dev->data;
 	struct tc_sm_t *tc = data->tc;
 
+#if !defined(CONFIG_TEST)
```
Why remove the assert when `CONFIG_TEST` is enabled? The point behind asserts is to alert programmer about conditions that must never be false but do become false due to incorrect code (e.g. invalid API usage).

sambhurst replied:
I agree that this is not the correct thing to do but twister was complaining:
```
/home/sam/Projects/github/zephyrproject/zephyr/include/zephyr/sys/util.h:109:45: error: invalid application of 'sizeof' to incomplete type 'const struct smf_state[]'
  109 |         ((size_t) (IS_ARRAY(array) + (sizeof(array) / sizeof((array)[0]))))
      |                                             ^
/home/sam/Projects/github/zephyrproject/zephyr/include/zephyr/sys/__assert.h:116:23: note: in definition of macro '__ASSERT'
  116 |                 if (!(test)) {                                            \
      |                       ^~~~
/home/sam/Projects/github/zephyrproject/zephyr/subsys/usb/usb_c/usbc_tc_common.c:130:26: note: in expansion of macro 'ARRAY_SIZE'
  130 |         __ASSERT(state < ARRAY_SIZE(tc_states), "invalid tc_state %d", state);
```
But the build works fine.

alevkoy commented on tests/subsys/usb_c/sink/dts/bindings/mock-tcpc.yaml:
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    Mock TypeC Port Controller used for testing
```
"Type-C"

alevkoy reviewed (commented):
Any chance you could split your last commit up into smaller chunks, e.g. the mock TCPC, the mock VBUS, the mock tester, and the test itself?

sambhurst commented:
> Any chance you could split your last commit up into smaller chunks, e.g. the mock TCPC, the mock VBUS, the mock tester, and the test itself?

Okay. I'll upload another PR with the mocks and tests split out into their own commits

sambhurst commented:
How odd. It seems like const struct smf_state[] should be completely defined regardless of CONFIG_TEST.

Is that newly happening in this PR?
@alevkoy This failure seemed odd to me and I was going to discuss it with @keith-zephyr. Yes, this is newly happening in the PR

sambhurst committed:
usb_c: Move state initialization structure to beginning of file

This change moves the state initialization structure to the
beginning of the file so the sizeof macro, used in __ASSERT,
can detect the type of the structure and determine its size.
Without this change, sizeof generates the following error:
"invalid application of 'sizeof' to incomplete type 'const struct
smf_state[]'"

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb_c: mocks: Add Mock TCPC for USB-C testing

The change adds a mock TCPC for USB-C testing. The mock
addes functionality needed by the USB-C subsystem to
detect a port partner, and send and receive PD messages.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb_c: mocks: Add Mock VBUS driver

This change addes a mock VBUS driver that adds functionality
used by the USB-C subsystem to measure and control VBUS.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb_c: mocks: Add Mock Tester for USB-C testing

This change addes a mock tester used For USB-C testing. Functionality
is added that enables the tests to act as a USB-C Port Partner.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

alevkoy reviewed (changes requested):
First pass. Got through the VBUS mock commit.

alevkoy commented on tests/subsys/usb_c/mocks/mock_tester.h:
```diff
@@ -0,0 +1,172 @@
+ *
+ * @retval true if UUT has a pending RX PD message
+ */
+bool tester_is_rx_msg_pending(void);
```
1. I find "rx" confusing here. The name of the function makes it seem like it's checking if the _tester_ has a pending received message. What about `tester_is_uut_msg_pending`?
2. But also, why is the test asking the _tester_ about the state of the _UUT_? If the TCPM provides an API that allows the tester to access this information, then couldn't the test just access it? Alternatively, if the tester is accessing it for its own purposes, then does this need to be part of the public API?

alevkoy commented on tests/subsys/usb_c/mocks/mock_tester.h:
```diff
@@ -0,0 +1,172 @@
+ *
+ * @retval message id
+ */
+int tester_get_msgid(void);
```
This is the message ID as understood by the tester, right? Is it the ID of a message transmitted by the tester or received by the tester? Is it the next one, or the previous one? These would be good things to note in the comments.

alevkoy commented on tests/subsys/usb_c/mocks/mock_tester.h:
```diff
@@ -0,0 +1,172 @@
+ *
+ * @param msg PD message sent from UUT
+ */
+void tester_get_uut_tx_data(struct pd_msg *msg);
```
I find "tx" confusing here, since it's not clear from which direction the message was transmitted. What about just `tester_get_utt_msg_data` (a _tester_ API for getting messages _from the UUT_)?

alevkoy commented on tests/subsys/usb_c/mocks/mock_tester.c:
```diff
@@ -0,0 +1,393 @@
+#include <zephyr/kernel.h>
+#include <string.h>
+#include <stddef.h>
+#include <zephyr/drivers/usb_c/usbc_pd.h>
```
I think you might need to go over this PR with `clang-format` to sort the headers.

alevkoy commented on tests/subsys/usb_c/mocks/mock_tester.c:
```diff
@@ -0,0 +1,393 @@
+#include "mock_tcpc.h"
+
+/**
+ * @breif Tester data
```
spelling, here and below

alevkoy commented on tests/subsys/usb_c/mocks/mock_tester.c:
```diff
@@ -0,0 +1,393 @@
+	/* pointer to mock tcpc */
+	struct mock_tcpc_data *tcpc;
+	/* CC1 voltage state */
+	enum tc_cc_voltage_state cc1;
```
Are these the voltages set by the UUT? If so, do they have different meaning than `uut_cc_pull`?

alevkoy commented on tests/subsys/usb_c/mocks/mock_tester.c:
```diff
@@ -0,0 +1,393 @@
+	struct pd_msg rx_msg;
+
+	/* PD message ID */
+	int msg_id;
```
Are these values the values for the tester or the UUT?

alevkoy commented on tests/subsys/usb_c/mocks/mock_tester.c:
```diff
@@ -0,0 +1,393 @@
+}
+
+/**
+ * @brief Get the PD message ID
```
These comments appear to duplicate the comments in the header file. I think you can remove them and just have comments for the public API in the header.

alevkoy commented on tests/subsys/usb_c/mocks/mock_tester.c:
```diff
@@ -0,0 +1,393 @@
+/**
+ * @brief Send a Data Message to the UUT
+ *
+ * @param type Data message type
```
What about `len`? (Nit: I would use something like `num_objects`. `tester.tx_msg.len` is evidently denominated in bytes, whereas `len` is denominated in 4-byte PD data objects.)

alevkoy commented on tests/subsys/usb_c/mocks/mock_tester.c:
```diff
@@ -0,0 +1,393 @@
+ *
+ * @retval CC2 voltage level
+ */
+enum tc_cc_voltage_state tester_get_cc2(void)
```
Nit: This function has the same logic as the previous function. You could have one helper function that just takes a CC value for CC1 or CC2 as a parameter and have these two functions be one-liners defined in terms of that function.

alevkoy commented on tests/subsys/usb_c/mocks/mock_tester.c:
```diff
@@ -0,0 +1,393 @@
+	enum tc_cc_polarity uut_polarity;
+
+	/* VBUS set by UUT or Tester */
+	int vbus;
```
How about `vbus_mv`?

alevkoy reviewed (commented):
Took another pass and got through the mock tester.

As a general comment, I had a hard time remembering the perspective of the different variables and functions, e.g. is this getting the CC value as seen by the tester or the UUT? I think naming changes could probably improve this.

alevkoy commented on tests/subsys/usb_c/sink/src/main.c:
```diff
@@ -0,0 +1,514 @@
+			/* Send Source Cap message */
+			tester_send_data_msg(PD_DATA_SOURCE_CAP, &pdo.raw_value, 1, true);
+			j++;
+			k = 0;
```
All of these single-character counters are a bit hard to follow. Can you unnest this a bit or comment more about the flow?

alevkoy commented on tests/subsys/usb_c/sink/src/main.c:
```diff
@@ -0,0 +1,514 @@
+	}
+
+	/* 7) The check fails if the UUT does not respond with a Request message. */
+	zassert_true(atomic_get(&data->uut_request), "UUT didn't send request message.");
```
This seems slightly incorrect:
1. `uut_request` gets set from the notify call from the stack to the DPM, whereas the result of `tester_get_uut_tx_data` is set by the tester when it actually receives the message. So you're taking the stack's word for it that it sent the message instead of having the tester verify that it received it.
2. `uut_request` gets set when the stack gets the RDO from the DPM, which it needs to do before it sends the Request message. So there will be at least some amount of time when `uut_request` is true, but the message hasn't been sent yet. In principle, the stack could request the RDO and never send the message, and the received message data would be invalid.

It seems like a more appropriate flow would be to poll the tester until it has received a message (step 6 above) and then check that that message is a Request and then check the RDO fields as below.

alevkoy commented on tests/subsys/usb_c/sink/src/main.c:
```diff
@@ -0,0 +1,514 @@
+	/* An explicit contract is now established. */
+
+	/*
+	 * 8) The Tester presents SinkTxOK if the test is in PD3 mode.
```
Is there a tester function that does this?

alevkoy commented on tests/subsys/usb_c/sink/src/main.c:
```diff
@@ -0,0 +1,514 @@
+	/* Send Accept and PS Ready messages to UUT */
+	uut_sent_request(dev);
+
+	zassert_equal(data->uut_received_src_cap_num, 1,
```
For these asserts, I don't have data consistency concerns as above, but I still have source-of-truth concerns. If this is modeled on a compliance test, then the tester should be the source of truth. Checking the results of these callbacks makes the stack the source of truth about its own behavior. To cleanly model a compliance test, it might make more sense to have a separate test (suite) focused on the DPM API and largely ignore it here.

If this _isn't_ supposed to model a compliance test, then I'm fine with checking this stuff.

alevkoy reviewed (commented):
Completed first pass on the whole PR
